### PR TITLE
Add covariance support and mutability feature for arrays and tuples

### PR DESCRIPTION
**Pull request by Anoukh**
## Purpose
This PR will add Covariance support for arrays and tuples. For example, the following variable declarations are now valid.
```ballerina
    Employee[] employeeArray;
    Person person1 = { name: "John" };
    Person[] personArray = employeeArray; // Valid since Person type is a super type of Employee type
    personArray[1] = person1; // Runtime exception since this will attempt to mutate the employeeArray

    Employee employee1 = { name: "John", permanent: true };
    (Employee, Employee) tuple1 = (employee1, employee1);
    (Employee, Person) tuple2 = tuple1; // Valid since Person type is a super type of Employee type
    tuple2[1] = person1; // Runtime exception since this will attempt to mutate tuple1 by inserting a Person type record
```

This PR depends on PR #9706 which unifies structural types

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 1]
```diff
@@ -1427,14 +1427,34 @@ private static void execStoreOpcodes(WorkerExecutionContext ctx, WorkerData sf,
...
+                        case TypeTags.ARRAY_TAG:
+                            BType elementType = ((BArrayType) list.getType()).getElementType();
+                            if (!checkCast(refReg, elementType)) {
+                                throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.TYPE_MISMATCH,
+                                        elementType, (refReg != null) ? refReg.getType() : BTypes.typeNull);
+                            }
+                            break;
+                        case TypeTags.TUPLE_TAG:
+                            elementType = ((BTupleType) list.getType()).getTupleTypes().get((int) index);
+                            if (!checkCast(refReg, elementType)) {
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: I think we bring this checkCast() check to the outside of (end of) the switch case, and make it commin for both the cases.
I think we will get here only for array and tuple types, so we can make that assumption, and ignore the default case (or make ArrayTag the default case).

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 2]
```diff
@@ -2508,6 +2528,62 @@ private static void handleVariableUnlock(WorkerExecutionContext ctx, BType[] typ
         }
     }
 
+    private static BRefType<?> execListGetOperation(BNewArray array, long index) {
+        switch (array.getTag()) {
+            case TypeTags.BOOLEAN_ARRAY_TAG:
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: We can use the TypeTag of the elementType of the array, instead of creating a new TypeTags set

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/TypeTags.java [hunk 3]
```diff
@@ -52,4 +52,11 @@
     public static final int FINITE_TYPE_TAG = 31;
     public static final int OBJECT_TYPE_TAG = 32;
     public static final int RECORD_TYPE_TAG = 33;
+    public static final int BOOLEAN_ARRAY_TAG = 34;
+    public static final int BYTE_ARRAY_TAG = 35;
+    public static final int FLOAT_ARRAY_TAG = 36;
+    public static final int INT_ARRAY_TAG = 37;
+    public static final int REF_ARRAY_TAG = 38;
+    public static final int STRING_ARRAY_TAG = 39;
```
Comment (SupunS) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/TypeTags.java]: We may don't need this. Refer the previous comment

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/lambda/FunctionPointersTest.java [hunk 4]
```diff
@@ -250,10 +250,4 @@ public void testFuncPointerConversion() {
         Assert.assertEquals(((BInteger) returns[0]).intValue(), 20);
     }
 
-    @Test(expectedExceptions = { BLangRuntimeException.class },
-            expectedExceptionsMessageRegExp = ".*error: error, message: 'function \\(Student\\) returns \\(int\\)' " +
-                    "cannot be cast to 'function \\(Person\\) returns \\(int\\)'.*")
-    public void testAnyToFuncPointerConversion_2() {
```
Comment (SupunS) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/lambda/FunctionPointersTest.java]: why is this removed?

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/ArrayMutabilityTest.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Test
+    public void testValidArrayAssignment() {
+        BValue[] returnValues = BRunUtil.invoke(compileResult, "testValidArrayAssignment");
+        Assert.assertTrue(((BBoolean) returnValues[0]).booleanValue(), "Expected value of true but found false");
+        Assert.assertEquals(((BInteger) returnValues[1]).intValue(), 100, "Expected value of 100");
+    }
+
+    @Test(description = "",
```
Comment (anupama-pathirage) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/ArrayMutabilityTest.java]: Shall we either remove this empty description or add meaningful description. Check all the methods

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/types/tuples/TupleMutabilityTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Test
+    public void testValidTupleAssignment() {
+        BValue[] returnValues = BRunUtil.invoke(compileResult, "testValidTupleAssignment");
+        Assert.assertTrue(((BBoolean) returnValues[0]).booleanValue(), "Expected value of true but found false");
+        Assert.assertEquals(((BInteger) returnValues[1]).intValue(), 100, "Expected value of 100");
+    }
+
+    @Test(description = "",
```
Comment (anupama-pathirage) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/types/tuples/TupleMutabilityTest.java]: Check the description

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 7]
```diff
@@ -3067,8 +3153,12 @@ private static boolean checkArrayCast(BType sourceType, BType targetType) {
...
 
             return checkArrayCast(sourceArrayType.getElementType(), targetArrayType.getElementType());
-        } else if (sourceType.getTag() == TypeTags.ARRAY_TAG) {
-            return targetType.getTag() == TypeTags.ANY_TAG;
+        } else if (targetType.getTag() == TypeTags.UNION_TAG) {
+            return checkUnionAssignable(sourceType, targetType);
+        }
+
+        if (targetType.getTag() == TypeTags.ANY_TAG) {
+            return true;
```
Comment (Anoukh) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: This should be first check

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 8]
```diff
@@ -3215,7 +3305,7 @@ public static boolean checkFunctionTypeEquality(BFunctionType source, BFunctionT
         }
 
         for (int i = 0; i < source.paramTypes.length; i++) {
-            if (!isSameType(source.paramTypes[i], target.paramTypes[i])) {
+            if (!isAssignable(source.paramTypes[i], target.paramTypes[i])) {
```
Comment (Anoukh) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Separate object equality from function type

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java (hunk 1 above)
Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Currently, `any` type is also entering this block. Will fix appropriate places so that only array and tuple will get here

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 9]
```diff
@@ -2509,6 +2517,69 @@ private static void handleVariableUnlock(WorkerExecutionContext ctx, BType[] typ
...
+    private static BRefType<?> execListGetOperation(BNewArray array, long index) {
+        if (array.getType().getTag() == TypeTags.ARRAY_TAG) {
+            switch (((BArrayType) array.getType()).getElementType().getTag()) {
+                case TypeTags.BOOLEAN_TAG:
+                    BBooleanArray bBooleanArray = (BBooleanArray) array;
+                    int i = bBooleanArray.get(index);
+                    return i == 0 ? new BBoolean(false) : new BBoolean(true);
+                case TypeTags.BYTE_TAG:
+                    BByteArray bByteArray = (BByteArray) array;
+                    return new BInteger(bByteArray.get(index));
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Should be BByte here

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 10]
```diff
@@ -2509,6 +2517,69 @@ private static void handleVariableUnlock(WorkerExecutionContext ctx, BType[] typ
         }
     }
 
+    private static BRefType<?> execListGetOperation(BNewArray array, long index) {
+        if (array.getType().getTag() == TypeTags.ARRAY_TAG) {
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Inverse this if and return the else fast. That will make the code look cleaner.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 11]
```diff
@@ -2509,6 +2517,69 @@ private static void handleVariableUnlock(WorkerExecutionContext ctx, BType[] typ
...
+                    return bRefValueArray.get(index);
+            }
+        }
+        // Else this is a tuple
+        BRefValueArray bRefValueArray = (BRefValueArray) array;
+        return bRefValueArray.get(index);
+    }
+
+    private static void execListAddOperation(BNewArray array, long index, BRefType refType) {
+        if (array.getType().getTag() == TypeTags.ARRAY_TAG) {
```
Comment (Kishanthan) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: Same here. Inverse the if with else.

File: bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java [hunk 12]
```diff
@@ -1426,14 +1426,23 @@ private static void execStoreOpcodes(WorkerExecutionContext ctx, WorkerData sf,
...
 
                 try {
-                    bArray.add(sf.longRegs[j], sf.refRegs[k]);
+                    long index = sf.longRegs[j];
+                    BRefType refReg = sf.refRegs[k];
+                    BType elementType = (list.getType().getTag() == TypeTags.ARRAY_TAG)
+                            ? ((BArrayType) list.getType()).getElementType()
+                            : ((BTupleType) list.getType()).getTupleTypes().get((int) index);
+                    if (!checkCast(refReg, elementType)) {
+                        throw BLangExceptionHelper.getRuntimeException(RuntimeErrors.INCOMPATIBLE_TYPE,
```
Comment (lankavitharana) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java]: I think you should handle the error here (like set the error to ctx and handle error)

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/ListUtils.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import org.ballerinalang.model.values.BInteger;
+import org.ballerinalang.model.values.BNewArray;
+import org.ballerinalang.model.values.BRefType;
+import org.ballerinalang.model.values.BRefValueArray;
+import org.ballerinalang.model.values.BString;
+import org.ballerinalang.model.values.BStringArray;
+
+/**
+ * Common utility methods used for List manipulation.
+ */
```
Comment (hasithaa) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/ListUtils.java]: Better to add version here.

Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/ListUtils.java]: Added in PR #10085
